File: app/src/main/java/xyz/fycz/myreader/model/third2/content/BookInfo.java
/*
 * This file is part of FYReader.
 *  FYReader is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  FYReader is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with FYReader.  If not, see <https://www.gnu.org/licenses/>.
 *
 *  Copyright (C) 2020 - 2022 fengyuecanzhu
 */

package xyz.fycz.myreader.model.third2.content;

import android.text.TextUtils;
import android.util.Log;


import io.reactivex.Observable;
import xyz.fycz.myreader.greendao.entity.Book;
import xyz.fycz.myreader.greendao.entity.rule.BookSource;
import xyz.fycz.myreader.greendao.entity.rule.InfoRule;
import xyz.fycz.myreader.model.third2.analyzeRule.AnalyzeByRegex;
import xyz.fycz.myreader.model.third2.analyzeRule.AnalyzeRule;
import xyz.fycz.myreader.util.utils.StringUtils;

import static android.text.TextUtils.isEmpty;

public class BookInfo {
    private String tag;
    private String sourceName;
    private BookSource source;

    public BookInfo(String tag, String sourceName, BookSource source) {
        this.tag = tag;
        this.sourceName = sourceName;
        this.source = source;
    }

    public Observable<Book> analyzeBookInfo(String s, final Book book) {
        return Observable.create(e -> {
            String baseUrl = book.getInfoUrl();

            InfoRule infoRule = source.getInfoRule();
            if (TextUtils.isEmpty(s)) {
                e.onError(new Throwable("书籍信息获取失败" + baseUrl));
                return;
            } else {
                Log.d(tag, "┌成功获取详情页");
                Log.d(tag, "└" + baseUrl);
            }
            book.setTag(tag);
            book.setSource(source.getSourceUrl());

            AnalyzeRule analyzer = new AnalyzeRule(book);
            analyzer.setContent(s, baseUrl);

            // 获取详情页预处理规则
            String ruleInfoInit = infoRule.getInit();
            boolean isRegex = false;
            if (!isEmpty(ruleInfoInit)) {
                // 仅使用java正则表达式提取书籍详情
                if (ruleInfoInit.startsWith(":")) {
                    isRegex = true;
                    ruleInfoInit = ruleInfoInit.substring(1);
                    Log.d(tag, "┌详情信息预处理");
                    AnalyzeByRegex.getInfoOfRegex(s, ruleInfoInit.split("&&"), 0, book, analyzer, source, tag);
                } else {
                    Object object = analyzer.getElement(ruleInfoInit);
                    if (object != null) {
                        analyzer.setContent(object);
                    }
                }
            }
            if (!isRegex) {
                Log.d(tag, "┌详情信息预处理");
                Object object = analyzer.getElement(ruleInfoInit);
                if (object != null) analyzer.setContent(object);
                Log.d(tag, "└详情预处理完成");

                Log.d(tag, "┌获取书名");
                String bookName = StringUtils.formatHtml(analyzer.getString(infoRule.getName()));
                if (!isEmpty(bookName)) book.setName(bookName);
                Log.d(tag, "└" + bookName);

                Log.d(tag, "┌获取作者");
                String bookAuthor = StringUtils.formatHtml(analyzer.getString(infoRule.getAuthor()));
                if (!isEmpty(bookAuthor)) book.setAuthor(bookAuthor);
                Log.d(tag, "└" + bookAuthor);

                Log.d(tag, "┌获取分类");
                String bookKind = analyzer.getString(infoRule.getType());
                Log.d(tag, "└" + bookKind);

                Log.d(tag, "┌获取最新章节");
                String bookLastChapter = analyzer.getString(infoRule.getLastChapter());
                if (!isEmpty(bookLastChapter)) book.setNewestChapterTitle(bookLastChapter);
                Log.d(tag, "└" + bookLastChapter);

                Log.d(tag, "┌获取简介");
                String bookIntroduce = analyzer.getString(infoRule.getDesc());
                if (!isEmpty(bookIntroduce)) book.setDesc(bookIntroduce);
                Log.d(tag, "└" + bookIntroduce);

                Log.d(tag, "┌获取封面");
                String bookCoverUrl = analyzer.getString(infoRule.getImgUrl(), true);
                if (!isEmpty(bookCoverUrl)) book.setImgUrl(bookCoverUrl);
                Log.d(tag, "└" + bookCoverUrl);

                Log.d(tag, "┌获取目录网址");
                String bookCatalogUrl = analyzer.getString(infoRule.getTocUrl(), true);
                if (isEmpty(bookCatalogUrl)) bookCatalogUrl = baseUrl;
                book.setChapterUrl(bookCatalogUrl);
                //如果目录页和详情页相同,暂存页面内容供获取目录用
                if (bookCatalogUrl.equals(baseUrl)) book.putCathe("ChapterListHtml", s);
                Log.d(tag, "└" + book.getChapterUrl());
                Log.d(tag, "-详情页解析完成");
            }
            e.onNext(book);
            e.onComplete();
        });
    }

}

File: app/src/main/java/xyz/fycz/myreader/model/third3/webBook/BookInfo.kt
/*
 * This file is part of FYReader.
 * FYReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FYReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with FYReader.  If not, see <https://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2020 - 2022 fengyuecanzhu
 */

package xyz.fycz.myreader.model.third3.webBook


import android.util.Log
import xyz.fycz.myreader.model.third3.analyzeRule.AnalyzeRule
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.ensureActive
import xyz.fycz.myreader.R
import xyz.fycz.myreader.application.App
import xyz.fycz.myreader.greendao.entity.Book
import xyz.fycz.myreader.greendao.entity.rule.BookSource
import xyz.fycz.myreader.model.third3.NoStackTraceException
import xyz.fycz.myreader.util.utils.HtmlFormatter
import xyz.fycz.myreader.util.utils.NetworkUtils

/**
 * 获取详情
 */
object BookInfo {

    @Throws(Exception::class)
    fun analyzeBookInfo(
        scope: CoroutineScope,
        bookSource: BookSource,
        book: Book,
        redirectUrl: String,
        baseUrl: String,
        body: String?,
        canReName: Boolean,
    ) {
        body ?: throw NoStackTraceException(
            App.getmContext().getString(R.string.error_get_web_content, baseUrl)
        )
        Log.d(bookSource.sourceUrl, "≡获取成功:${baseUrl}")
        Log.d(bookSource.sourceUrl, body)
        val analyzeRule = AnalyzeRule(book, bookSource)
        analyzeRule.setContent(body).setBaseUrl(baseUrl)
        analyzeRule.setRedirectUrl(redirectUrl)
        analyzeBookInfo(scope, book, body, analyzeRule, bookSource, baseUrl, redirectUrl, canReName)
    }

    fun analyzeBookInfo(
        scope: CoroutineScope,
        book: Book,
        body: String,
        analyzeRule: AnalyzeRule,
        bookSource: BookSource,
        baseUrl: String,
        redirectUrl: String,
        canReName: Boolean,
    ) {
        val infoRule = bookSource.infoRule
        infoRule.init?.let {
            if (it.isNotBlank()) {
                scope.ensureActive()
                Log.d(bookSource.sourceUrl, "≡执行详情页初始化规则")
                analyzeRule.setContent(analyzeRule.getElement(it))
            }
        }
        //val mCanReName = canReName && !infoRule.canReName.isNullOrBlank()
        val mCanReName = false
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取书名")
        BookList.formatBookName(analyzeRule.getString(infoRule.name)).let {
            //if (it.isNotEmpty() && (mCanReName || book.name.isEmpty())) {
            if (it.isNotEmpty()) {
                book.name = it
            }
            Log.d(bookSource.sourceUrl, "└${it}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取作者")
        BookList.formatBookAuthor(analyzeRule.getString(infoRule.author)).let {
            //if (it.isNotEmpty() && (mCanReName || book.author.isEmpty())) {
            if (it.isNotEmpty()) {
                book.author = it
            }
            Log.d(bookSource.sourceUrl, "└${it}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取分类")
        try {
            analyzeRule.getStringList(infoRule.type)
                ?.joinToString(",")
                ?.let {
                    if (it.isNotEmpty()) book.type = it
                }
            Log.d(bookSource.sourceUrl, "└${book.type}")
        } catch (e: Exception) {
            Log.d(bookSource.sourceUrl, "└${e.localizedMessage}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取字数")
        try {
            /*wordCountFormat(analyzeRule.getString(infoRule.wordCount)).let {
                if (it.isNotEmpty()) book.wordCount = it
            }*/
            analyzeRule.getString(infoRule.wordCount).let {
                if (it.isNotEmpty()) book.wordCount = it
            }
            Log.d(bookSource.sourceUrl, "└${book.wordCount}")
        } catch (e: Exception) {
            Log.d(bookSource.sourceUrl, "└${e.localizedMessage}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取最新章节")
        try {
            analyzeRule.getString(infoRule.lastChapter).let {
                if (it.isNotEmpty()) book.newestChapterTitle = it
            }
            Log.d(bookSource.sourceUrl, "└${book.newestChapterTitle}")
        } catch (e: Exception) {
            Log.d(bookSource.sourceUrl, "└${e.localizedMessage}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取简介")
        try {
            analyzeRule.getString(infoRule.desc).let {
                if (it.isNotEmpty()) book.desc = HtmlFormatter.format(it)
            }
            Log.d(bookSource.sourceUrl, "└${book.desc}")
        } catch (e: Exception) {
            Log.d(bookSource.sourceUrl, "└${e.localizedMessage}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取封面链接")
        try {
            analyzeRule.getString(infoRule.imgUrl).let {
                if (it.isNotEmpty()) book.imgUrl = NetworkUtils.getAbsoluteURL(baseUrl, it)
            }
            Log.d(bookSource.sourceUrl, "└${book.imgUrl}")
        } catch (e: Exception) {
            Log.d(bookSource.sourceUrl, "└${e.localizedMessage}")
        }
        scope.ensureActive()
        Log.d(bookSource.sourceUrl, "┌获取目录链接")
        book.chapterUrl = analyzeRule.getString(infoRule.tocUrl, isUrl = true)
        if (book.chapterUrl.isEmpty()) book.chapterUrl = redirectUrl
        if (book.chapterUrl == redirectUrl) {
            book.putCathe("tocHtml", body)
            //book.tocHtml = body
        }
        Log.d(bookSource.sourceUrl, "└${book.chapterUrl}")
    }

}